protocol_rest.go

```diff
@@ -140,7 +140,12 @@ func (r restClientProtocol) requestNeedsPrep(op *operation) bool {
 	return len(op.restTarget.vars) != 0 ||
 		len(op.request.URL.Query()) != 0 ||
 		op.restTarget.requestBodyFields != nil ||
-		restHTTPBodyRequest(op)
+		restHTTPBodyRequest(op) ||
+		restHTTPBodyRequestIsEmpty(op)
+}
+
+func restHTTPBodyRequestIsEmpty(op *operation) bool {
+	return op.request.ContentLength < 0
```

Comment:
This is checking for an unknown `ContentLength` as `-1` is valid for a non empty body where the content-length can't be deduced up front.

Reply:
I added some unit tests and took another stab at it using the restTarget values. 

I'm open to other fixes.

Reply:
I believe the issue is here: https://github.com/connectrpc/vanguard-go/blob/7aae240d504a5f2ce8aa6cf148232fcbf88eaf36/transcoder.go#L2087
We need to not fallthrough to the default case as an empty body is not a valid JSON request. We must always use the `prepareUnmarshalRequest` for decoding the message if needed. We also must always emit a request message when we hit EOF here:
https://github.com/connectrpc/vanguard-go/blob/7aae240d504a5f2ce8aa6cf148232fcbf88eaf36/transcoder.go#L952-L953

I think for now the simplest fix is for any GET request to return true on `requestNeedsPrep`. So this would add  a simple `op.request.Method == http.MethodGET` to the logic . The `prepareUnmarshaledRequest` is close to a no-op so shouldn't add much overhead. And almost all `GET` requests will require preparation of the request message.